pkg/router/template/router.go

```diff
@@ -311,31 +323,171 @@ func (r *templateRouter) readState() error {
 	return json.Unmarshal(data, &r.state)
 }
 
-// Commit applies the changes made to the router configuration - persists
+// Define a type for the pluggable commit function
+type CommitFunc func() error
+
+// This should ONLY be used for external testing hooks
+func (r *templateRouter) SetCommitFunc(commitFunc CommitFunc) {
+	r.commitFunc = commitFunc
+}
+
+// Commit calls the realCommit worker with a flag saying that it has
+// synced.  This should be called only when the router state is
+// consistent.  realCommit() below can be called on initial router
+// load to load the haproxy without a full state sync.
+func (r *templateRouter) Commit() {
+	r.realCommit(true)
+}
+
+// realCommit applies the changes made to the router configuration - persists
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
+//
+// The hasSynced argument specifies whether the state is fully synchronized.
+//
 // Note: If this is changed FakeCommit() in fake.go should also be updated
-func (r *templateRouter) Commit() {
+func (r *templateRouter) realCommit(hasSynced bool) {
 	r.lock.Lock()
 
-	if !r.synced {
-		glog.V(4).Infof("Router state synchronized for the first time")
-		r.synced = true
-		r.stateChanged = true
+	if hasSynced {
+		// Only update our sync state if we've really synced.  We want
+		// to let the router start early so that it can return health
+		// while the state synchronizes, but before the main ports are
+		// bound
+		if !r.synced {
+			glog.V(4).Infof("Router state synchronized for the first time, allowing rapid reload")
+			r.synced = true
+			r.stateChanged = true
+
+			// Clean out the last reload variables so the reload from the first sync doesn't have to wait
+			r.lastReloadStart = time.Time{}
+			r.lastReloadEnd = time.Time{}
+		}
+	}
+
+	if r.stateChanged {
+		// If the state changed, then we need to commit
+		glog.V(8).Infof("Commit called and the state has changed, could reload")
+
+		if r.commitReqTime == nil {
+			// There is no scheduled commit worker, so set the time we started and
+			// invoke the worker code.  It will decide if it can run now, or if it
+			// needs to schedule a callback.
+			//
+			// We need to track the earliest commit time so we can do burst supression.
+			now := time.Now()
+			r.commitReqTime = &now
+			r.lock.Unlock()
```

Comment:
What if you just `defer r.lock.Unlock()` at the top, and then in this block do `defer r.commitWorker()` so the locking is a bit simpler?  The second defer should run after the unlock.

Reply:
We can't do it that way because the unlock would be called after the commitWorker because defers are called in reverse order from definition.